FV3/coarse_graining/coarse_graining.F90

```diff
@@ -582,21 +585,45 @@ subroutine vertical_remapping_requirements(delp, area, ptop, phalf, upsampled_co
     deallocate(coarse_phalf)
    end subroutine vertical_remapping_requirements
 
-   subroutine mask_area_weights(area, phalf, upsampled_coarse_phalf, masked_area_weights)
+   subroutine mask_area_weights(area, phalf, upsampled_coarse_phalf, extrapolate, masked_area_weights)
     real, intent(in) :: area(is:ie,js:je)
     real, intent(in) :: phalf(is:ie,js:je,1:npz+1)
     real, intent(in) :: upsampled_coarse_phalf(is:ie,js:je,1:npz+1)
+    logical, intent(in) :: extrapolate
     real, intent(out) :: masked_area_weights(is:ie,js:je,1:npz)
 
+    real, allocatable :: upsampled_coarse_pfull(:,:,:)
     integer :: k
 
-    do k = 1, npz
-      where (upsampled_coarse_phalf(is:ie,js:je,k+1) .lt. phalf(is:ie,js:je,npz+1))
-        masked_area_weights(is:ie,js:je,k) = area(is:ie,js:je)
-      elsewhere
-        masked_area_weights(is:ie,js:je,k) = 0.0
-      endwhere
-    enddo
+    ! Even in "extrapolation" mode we extrapolate in a limited sense.  We use
+    ! nearest neighbor extrapolation in the event that the pressure at the
+    ! vertical midpoint of the cell is less than that of the surface pressure
+    ! in the fine-grid column; otherwise do not attempt to extrapolate and
+    ! mask the points when computing horizontal averages.  This approach
+    ! allows us to generally use all grid cells in the lowest model level
+    ! over ocean, while maintaining masking in land regions with variable
+    ! topography; our stricter no-extrapolation approach often masks 50% or so
+    ! of the grid cells there.
+    if (extrapolate) then
+      allocate(upsampled_coarse_pfull(is:ie,js:je,1:npz))
+      call compute_pfull_from_phalf(upsampled_coarse_phalf, upsampled_coarse_pfull)
+
+      do k = 1, npz
+         where (upsampled_coarse_pfull(is:ie,js:je,k) .lt. phalf(is:ie,js:je,npz+1))
```

Comment:
Indeed I think it's somewhat arbitrary since this condition is only meant to provide an approximate buffer within which to allow extrapolation.  I think I'll opt to keep things with `<` rather than `<=` for symmetry with the no-extrapolation case.